src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/CollectTraceActionTests.cs

```diff
@@ -0,0 +1,159 @@
+
+                rootOptions.CreateCollectionRule(DefaultRuleName)
+                    .AddCollectTraceAction(ExpectedProviders, ExpectedEgressProvider, out CollectTraceOptions collectTraceOptions)
+                    .SetStartupTrigger();
+
+                collectTraceOptions.Duration = TimeSpan.FromSeconds(2);
+            }, async host =>
+            {
+                await PerformTrace(host);
+            });
+        }
+
+        private async Task PerformTrace(IHost host)
+        {
+            IOptionsMonitor<CollectionRuleOptions> ruleOptionsMonitor = host.Services.GetService<IOptionsMonitor<CollectionRuleOptions>>();
+            CollectTraceOptions options = (CollectTraceOptions)ruleOptionsMonitor.Get(DefaultRuleName).Actions[0].Settings;
+
+            ICollectionRuleActionFactoryProxy factory;
+            Assert.True(host.Services.GetService<ICollectionRuleActionOperations>().TryCreateFactory(KnownCollectionRuleActions.CollectTrace, out factory));
+
+            EndpointInfoSourceCallback callback = new(_outputHelper);
+            await using var source = _endpointUtilities.CreateServerSource(out string transportName, callback);
+            source.Start();
+
+            AppRunner runner = _endpointUtilities.CreateAppRunner(transportName, TargetFrameworkMoniker.Net60); // Arbitrarily chose Net60
+
+            Task<IEndpointInfo> newEndpointInfoTask = callback.WaitForNewEndpointInfoAsync(runner, CommonTestTimeouts.StartProcess);
+
+            await runner.ExecuteAsync(async () =>
+            {
+                IEndpointInfo endpointInfo = await newEndpointInfoTask;
+
+                ICollectionRuleAction action = factory.Create(endpointInfo, options);
+
+                using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(CommonTestTimeouts.TraceTimeout);
+
+                CollectionRuleActionResult result;
+                try
+                {
+                    await action.StartAsync(cancellationTokenSource.Token);
+
+                    result = await action.WaitForCompletionAsync(cancellationTokenSource.Token);
+                }
+                finally
+                {
+                    await DisposableHelper.DisposeAsync(action);
+                }
+
+                Assert.NotNull(result.OutputValues);
+                Assert.True(result.OutputValues.TryGetValue(CollectionRuleActionConstants.EgressPathOutputValueName, out string egressPath));
+                Assert.True(File.Exists(egressPath));
+
+                using FileStream traceStream = new(egressPath, FileMode.Open, FileAccess.Read);
+                Assert.NotNull(traceStream);
+
+                await ValidateTrace(traceStream);
+
+                await runner.SendCommandAsync(TestAppScenarios.AsyncWait.Commands.Continue);
+            });
+        }
+
+        private static async Task ValidateTrace(Stream traceStream)
```

Comment:
The better way to do this is to use an EventPipeEventSource and verify that it can produce some event callback. See https://github.com/dotnet/diagnostics/blob/d425906a41dc8e2df0cf22766716feebe4634a5d/src/tests/Microsoft.Diagnostics.Monitoring.EventPipe/EventTracePipelineUnitTests.cs#L54 as an example.

Reply:
I'm not entirely sure what to do here, so I'm going to try and clarify a few quick things. Also, if my understanding in 2 is completely off, just let me know :)

1. Do we want to completely take out the existing check for the headers, or are we supplementing it with this?
2. My understanding is that in the code you linked, line 61 gets run when there's an event callback (which is what we want to verify exists); do we just want to assert that we hit that line (potentially setting a boolean to true there that gets checked outside the block), or are we actually doing any validation on the objects in the trace itself?

Reply:
1. Whole replacement. EventPipeEventSource will check that the format is as expected.
2. Set a boolean that any event was captured is fine and then assert that is is true after the Process call returns.

In the future, we can verify that we got the expected types of events if we really feel that is necessary.